go/libraries/doltcore/schema/index_coll.go

```diff
@@ -0,0 +1,349 @@
+	ReferencesTag(tag uint64) []Index
+	// RemoveIndex removes an index from the table metadata.
+	RemoveIndex(indexName string) (Index, error)
+	// RenameIndex renames an index in the table metadata.
+	RenameIndex(oldName, newName string) (Index, error)
+}
+
+type IndexTagChange struct {
+	oldTag uint64
+	newTag uint64
+}
+
+func NewIndexTagChange(oldTag uint64, newTag uint64) IndexTagChange {
+	return IndexTagChange{
+		oldTag: oldTag,
+		newTag: newTag,
+	}
+}
+
+type indexCollectionImpl struct {
+	colColl       *ColCollection
+	indexes       map[string]*indexImpl
+	colTagToIndex map[uint64][]*indexImpl
+	pks           []uint64
+}
+
+func NewIndexCollection(cols *ColCollection) IndexCollection {
+	ixc := &indexCollectionImpl{
+		colColl:       cols,
+		indexes:       make(map[string]*indexImpl),
+		colTagToIndex: make(map[uint64][]*indexImpl),
+	}
+	if cols != nil {
+		for _, col := range cols.cols {
+			ixc.colTagToIndex[col.Tag] = nil
+			if col.IsPartOfPK {
+				ixc.pks = append(ixc.pks, col.Tag)
+			}
+		}
+	}
+	return ixc
+}
+
+func (ixc *indexCollectionImpl) AddIndex(indexes ...Index) {
+	for _, indexInterface := range indexes {
+		index, ok := indexInterface.(*indexImpl)
+		if !ok {
+			panic(fmt.Errorf("unknown index type: %T", indexInterface))
+		}
+		index = index.copy()
+		index.indexColl = ixc
+		oldNamedIndex, ok := ixc.indexes[index.name]
+		if ok {
+			ixc.removeIndex(oldNamedIndex)
+		}
+		oldTaggedIndex := ixc.containsColumnTagCollection(index.tags...)
+		if oldTaggedIndex != nil {
+			ixc.removeIndex(oldTaggedIndex)
+		}
+		ixc.indexes[index.name] = index
+		for _, tag := range index.tags {
+			ixc.colTagToIndex[tag] = append(ixc.colTagToIndex[tag], index)
+		}
+	}
+}
+
+func (ixc *indexCollectionImpl) AddIndexByColNames(indexName string, cols []string, isUnique bool, comment string) (Index, error) {
+	if ixc.Contains(indexName) {
+		return nil, fmt.Errorf("`%s` already exists as an index for this table", indexName)
+	}
+	if ixc.ContainsColumnCollection(cols...) {
+		return nil, fmt.Errorf("cannot create a duplicate index on this table")
+	}
+	index := &indexImpl{
+		indexColl: ixc,
+		name:      indexName,
+		isUnique:  isUnique,
+		comment:   comment,
+	}
+	ixc.indexes[indexName] = index
+	tags := make([]uint64, len(cols))
+	for i, col := range cols {
+		tag := ixc.colColl.NameToCol[col].Tag
+		tags[i] = tag
+		ixc.colTagToIndex[tag] = append(ixc.colTagToIndex[tag], index)
+	}
+	index.tags = tags
+	index.allTags = combineAllTags(tags, ixc.pks)
+	return index, nil
+}
+
+func (ixc *indexCollectionImpl) AddIndexByColTags(indexName string, tags []uint64, isUnique bool, comment string) (Index, error) {
+	if ixc.Contains(indexName) {
+		return nil, fmt.Errorf("`%s` already exists as an index for this table", indexName)
+	}
+	if ixc.ContainsColumnTagCollection(tags...) {
+		return nil, fmt.Errorf("cannot create a duplicate index on this table")
+	}
+	index := &indexImpl{
```

Comment:
A little different than above, this is only called when unmarshalling, so it's always going to have the tags, but I should still check here since that's not guaranteed to always be true.